### DISCUSSION TITLE
Question: What is `maximum_sinking_velocity(bgc)` expected behaviour?

### DISCUSSION THREAD

**Discussion by radka-j** [OceanBioME/OceanBioME.jl#314] [Q&A]
Hi! 

I'm working on implementing a biogeochemistry model using OceanBioME and have found it incredibly useful, so thank you! 

I have a question about the`maximum_sinking_velocity(bgc)` method. I have noticed that `NPZD` and `LOBSTER` both define it but not `PISCES`. it is also implemented in the main `Biogeochemistry` class where it is set to return 0 and this is the method that seems to be called in the [code](https://github.com/OceanBioME/OceanBioME.jl/blob/main/src/Utils/timestep.jl). 

Additionally, if I use the example in the README and try to call this method, it works if called on the `biogeochemistry` class (returning a 0) but gives an error if called on `biogeochemistry.underlying_biogeochemistry` (which is `NPZD` in this case):

```
julia> OceanBioME.maximum_sinking_velocity(biogeochemistry)
0.0
```

```
julia> OceanBioME.maximum_sinking_velocity(biogeochemistry.underlying_biogeochemistry)
ERROR: type Field has no field w
```

Is the expected behaviour that the returned value is always 0? What is this used for?

**Comment by jagoosw** [comment]
Hi @radka-j,

The intention of this function was always to return the maximum sinking velocity that a bgc model imposed, but it seems like it was never tested and then forgotten. They could then be used by the code you linked in a timestep wizard-like:
https://github.com/OceanBioME/OceanBioME.jl/blob/eb89012976f33bc50814ee6a297d2b6cdecdd32a/paper/figures/column.jl#L63

`maximum_sinking_velocity(bgc)` is hitting the fallback:
https://github.com/OceanBioME/OceanBioME.jl/blob/eb89012976f33bc50814ee6a297d2b6cdecdd32a/src/OceanBioME.jl#L163
because I never implemented:
```julia
maximum_sinking_velocity(bgc::CompleteBiogeochemistry) = maximum_sinking_velocity(bgc.underlying_biogeochemistry)
```

and at some point I forgot to update the NPZD and LOBSTER methods which should be like:
```julia
 @inline maximum_sinking_velocity(bgc::LOBSTER) = maximum(abs, bgc.sinking_velocities.bPOM) 
```
not
https://github.com/OceanBioME/OceanBioME.jl/blob/eb89012976f33bc50814ee6a297d2b6cdecdd32a/src/Models/AdvectedPopulations/LOBSTER/LOBSTER.jl#L442

I can fix this at some point, or if you'd like to use it feel free to PR the changes if I don't get round to it soon enough!

**Comment by glwagner** [comment]
> The intention of this function was always to return the maximum sinking velocity that a bgc model imposed, but it seems like it was never tested and then forgotten. They could then be used by the code you linked in a timestep wizard-like:

That'd be appropriate to implement in Oceananigans if it is extending `TimeStepWizard`.

**Comment by jagoosw** [comment]
> > The intention of this function was always to return the maximum sinking velocity that a bgc model imposed, but it seems like it was never tested and then forgotten. They could then be used by the code you linked in a timestep wizard-like:
> 
> That'd be appropriate to implement in Oceananigans if it is extending `TimeStepWizard`.

That's probably a much better approach since it should just be able to return something like `maximum(map(n->maximum(drift_velocities[n].w), 1:length(drift_velocities))`, I guess this never actually needed to be defined for all models but we can just add methods to make it more efficient if we want to

**Comment by glwagner** [comment]
I think the right solution is to use `total_velocities` to compute the CFL here:

https://github.com/CliMA/Oceananigans.jl/blob/5dd01ddfe73dd03515b52e6499b27c1b61c71364/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl#L259-L261

This is guaranteed to work universally I think and wouldn't require maintaining any code in OceanBioME specifically?
